chartpress.py

```diff
@@ -167,7 +172,6 @@ def image_needs_pushing(image):
         return False
 
 
-@lru_cache()
```

Comment:
image_needs_building() changed during the duration of testing leading to an outdated value were used.

Reply:
You can clear the cache by calling

```python
image_needs_rebuilding.cache_clear()
```

in the test fixture. That way, we don't lose this. We should also put this decorator on `latest_tag_or_mod_commit`.

Reply:
None of these functions are pure though, hmmm... Do we really want to cache them at all?

I think the key question is: will chartpress, except for within a testing environment, ever risk reusing a cached value that no longer is relevant?

`latest_tag_or_mod_commit` can become outdated by git operations. No such operations are expected to be made during chartpress start and exit when used as a CLI, which is the duration of the cache as far as I understand. This function is currently called once for each chart's path, and once for each chart's images paths.

`image_needs_building` can become outdated by a docker build operation for example. It is called once for each image that is meant to be built from the build_images function.

`image_needs_pushing` can become outdated by a docker push operation for example, and is called TWICE, once to from image_needs_building and once if the `--push` flag is passed to verify it needs to be pushed (... and the tag isn't set? Adding tag implies `--force` to push and build currently. Hmm...)

---

What is your take considering this, should we cache? Also, can you confirm if I got it right that lru_cache will be reset in between calls to the chartpress CLI, where the process is shut down, but perhaps not if you have one process running that in turn has imported chartpress and use the functions directly without shutting down?

---

Do you think @lru_cache usage may relate to why travis-ci mess up with https://github.com/jupyterhub/chartpress/pull/68#issuecomment-545582996 ?

Reply:
(I updated the comment above to exclude no longer relevant discussion)

Reply:
While everything currently works, the tests will be a lot harder to get right and maintain with caching in place. I spent some extra hours on this but it doesn't doesn't do anything for the CLI except for image_needs_pushing cache in the specific circumstance that: an image was available locally, but not remotely, and we used `--push`.

Oh hmmmm, I can make a call to cache_clear alongside every call to chartpress during testing, it would make the test properly minimic the CLI which we want! I'm quite happy with this path as well!

Reply:
I consider this resolved in ac85a6e